### PR TITLE
[model, config] feat: add MiniMax H3 model support

### PR DESCRIPTION
**Pull request by vvyuervv**
### What does this PR do?

> This PR adds MiniMax H3 model support to VeOmni, including the full model implementation, training/inference pipeline integration, and related utilities, with full functionality validated and adapted for the Ascend environment.                                                                                                                                                                                                                                                    

 - MiniMax H3 model module (veomni/models/diffusers/minimax_h3/): complete implementation with                                                                                                                 
     - Core components: audio VAE, video VAE, DiT, text encoder, packed sequence, offline loader, flow-match scheduler                                                                                         
     - Condition and transformer sub-modules (configurations + modelings)                                                                                                                                      
     - Inference (inference.py, ~1.3k lines) entry points                                                                                                                                                                                                                                                    
                                                                                                                                        
 - Configs & examples: training configs (configs/dit/minimax_h3_*.yaml), inference task (tasks/infer/infer_minimax_h3.py), usage docs                                                       

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test
> The total loss curve of FL2VA task training shows that the loss tends to converge.

<img width="1650" height="1950" alt="loss_curves" src="https://github.com/user-attachments/assets/7bc0cde4-e710-4604-a70f-d7ba7b580789" />

> Inference Videos of T2VA and FL2VA Tasks

T2VA：
https://github.com/user-attachments/assets/e5635cf4-4c7a-4fbe-b7e1-2ca5647e9484

FL2VA ：
https://github.com/user-attachments/assets/46930000-3891-4045-bab1-6a6ae818075a



### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

> High-level design description and specific change list.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added MiniMax H3 support for synchronized text-to-video and text-to-audio generation.
  * Added keyframe-conditioned video and audio generation with MP4 output.
  * Added online and offline multimodal dataset processing.
  * Added offline embedding and DiT training configurations.
  * Added configurable AdamW beta parameters and enhanced training options.

* **Documentation**
  * Added a MiniMax H3 quick-start guide covering setup, data preparation, training, and inference.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: veomni/utils/audio.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import torch
```
Comment (Coach257) [inline veomni/utils/audio.py]: Please refer to veomni/data/multimodal/audio_utils for audio utils.

File: veomni/utils/audio_video.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+from fractions import Fraction
```
Comment (Coach257) [inline veomni/utils/audio_video.py]: Please refer to veomni/data/multimodal/video_utils.py

File: veomni/models/diffusers/minimax_h3/minimax_h3_condition/modeling_minimax_h3_condition.py [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+        # Processor / tokenizer
+        try:
+            from transformers import AutoProcessor
+
+            proc_path = f"{base}/{config.processor_subfolder}" if base else None
+            if proc_path:
+                self._processor = AutoProcessor.from_pretrained(proc_path)
+                self._tokenizer = self._processor.tokenizer
+        except Exception:
+            logger.warning("Failed to load Qwen3-VL processor from %s", proc_path)
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/minimax_h3_condition/modeling_minimax_h3_condition.py]: Fixed

File: veomni/models/diffusers/minimax_h3/minimax_h3_condition/modeling_minimax_h3_condition.py [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+                    from PIL import Image
+
+                    keyframe_images = []
+                    for idx in cfg.keyframe_indices:
+                        frame_tensor = video_frames[idx]
+                        if frame_tensor.dim() == 3:
+                            frame_np = (frame_tensor.permute(1, 2, 0).cpu().numpy() * 255).astype("uint8")
+                            keyframe_images.append(Image.fromarray(frame_np))
+            else:
+                keyframe_images = None
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/minimax_h3_condition/modeling_minimax_h3_condition.py:282]: fixed

File: veomni/utils/audio_video.py (hunk 2 above)
Comment (vvyuervv) [reply veomni/utils/audio_video.py]: The required related functions have been merged into audio_utils and video_utils, and the newly added files under utils have been deleted.

File: veomni/data/multimodal/dit/data_transform.py [hunk 5]
```diff
@@ -20,9 +22,112 @@ def process_dit_online_example(example, source_name, **kwargs):
...
 @DATA_TRANSFORM_REGISTRY.register("dit_offline")
 def process_dit_offline_example(example, **kwargs):
     import pickle as pk
 
     processed_example = {key: pk.loads(value) for key, value in example.items()}
     return [processed_example]
+
+
+@DATA_TRANSFORM_REGISTRY.register("minimax_h3_offline")
+def process_minimax_h3_offline_example(example, **kwargs):
```
Comment (Coach257) [inline veomni/data/multimodal/dit/data_transform.py]: Directly use `process_dit_offline_example`

File: veomni/trainer/base.py [hunk 6]
```diff
@@ -93,6 +94,11 @@
 
 logger = logging.get_logger(__name__)
 
+# Trainer registry keyed by data.datasets_type. Task entry points look up the
+# trainer class here (falling back to the default), letting per-model packages
+# ship their own trainer subclasses (see models/diffusers/minimax_h3/trainer.py).
+TRAINER_REGISTRY = Registry("Trainer")
```
Comment (Coach257) [inline veomni/trainer/base.py]: We do not use trainer_registry currently. The customed trainer diffs in optmizer & lr_scheduler. Maybe add optimizer_type or scheduler_type instead of building a new trainer.

File: veomni/data/multimodal/dit/minimax_h3_preprocess.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+condition_model.get_condition() — not provided by the preprocessor.
+"""
+
+import os
+
+from ..preprocess import PREPROCESSOR_REGISTRY
+
+
+@PREPROCESSOR_REGISTRY.register("minimax_h3")
+def minimax_h3_preprocess(conversations, **kwargs):
```
Comment (Coach257) [inline veomni/data/multimodal/dit/minimax_h3_preprocess.py]: try to extend `process_dit_online_example` to audio process.
```
processed_example = {
        "inputs": inputs,
        "outputs": outputs,
        "images": images,
        "videos": videos,
        "audios": audios,
}
```
`prompt` is `inputs`, `videos` could be dict `{video:, audio:}`

File: docs/examples/minimax_h3.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
+# MiniMax H3 FL2VA 快速开始
```
Comment (Coach257) [inline docs/examples/minimax_h3.md]: Docs in English.

File: veomni/data/multimodal/dit/data_transform.py (hunk 5 above)
Comment (vvyuervv) [reply veomni/data/multimodal/dit/data_transform.py]: done

File: veomni/trainer/base.py (hunk 6 above)
Comment (vvyuervv) [reply veomni/trainer/base.py]: done

File: docs/examples/minimax_h3.md (hunk 8 above)
Comment (vvyuervv) [reply docs/examples/minimax_h3.md]: done

File: veomni/data/multimodal/audio_utils.py [hunk 9]
```diff
@@ -200,3 +202,86 @@ def save_audio_tensor_to_file(
...
+    for rframe in audio_resampler.resample(frame_in):
+        if rframe.pts is None:
+            rframe.pts = audio_next_pts
+        audio_next_pts += rframe.samples
+        rframe.sample_rate = frame_in.sample_rate
+        container.mux(audio_stream.encode(rframe))
+
+    # flush audio encoder
+    for packet in audio_stream.encode():
+        container.mux(packet)
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/data/multimodal/audio_utils.py:251]: This issue has been resolved in commit id 9941f1b8.

File: veomni/data/multimodal/video_utils.py [hunk 10]
```diff
@@ -873,3 +876,61 @@ def save_video_tensors_to_file(
...
+    duration = len(video) / fps
+    if audio_sample_rate is None:
+        audio_sample_rate = int(audio.shape[-1] / duration)
+
+    width, height = video[0].size
+    container = av.open(output_path, mode="w")
+    stream = container.add_stream("libx264", rate=int(fps))
+    stream.width = width
+    stream.height = height
+    stream.pix_fmt = "yuv420p"
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/data/multimodal/video_utils.py:923]: This issue has been resolved in commit id 9941f1b8.

File: veomni/models/diffusers/minimax_h3/minimax_h3_core/minimax_h3_video_vae.py [hunk 11]
```diff
@@ -0,0 +1,842 @@
...
+                x = x.transpose(1, 2)
+            B, _, T, _, _ = x.shape
+            x = rearrange(
+                self.transform(rearrange(x, "b c t h w -> (b t) c h w")), "(b t) c h w -> b c t h w", b=B, t=T
+            )
+        else:
+            if x.ndim == 4 and x.shape[1] == 3:
+                x = x.transpose(0, 1) if x.shape[0] == 3 else x
+            x = self.transform(x)
+        return x.contiguous()
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/minimax_h3_core/minimax_h3_video_vae.py:471]: This issue has been resolved in commit id 9941f1b8.

File: veomni/models/diffusers/minimax_h3/inference.py [hunk 12]
```diff
@@ -0,0 +1,1345 @@
...
+        negative_prompt: str = " ",
+        height: int = 768,
+        width: int = 1344,
+        num_frames: int = 124,
+        num_inference_steps: int = 50,
+        seed: int = 42,
+        rand_device: str = "cpu",
+        cfg_scale: float = 1.0,
+        flow_shift: float = 12.0,
+        audio_flow_shift: float = 3.0,
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/inference.py]: This issue has been resolved in commit id 9941f1b8.

File: veomni/models/diffusers/minimax_h3/inference.py [hunk 13]
```diff
@@ -0,0 +1,1345 @@
...
+
+        return {
+            "img_pos": img_pos,
+            "audio_pos": audio_pos,
+            "text_pos": torch.arange(0, text_len),
+            "img_position_ids": g[None],
+            "token_tags": token_tags,
+            "cu_seqlens": torch.tensor([0, used, seq_len], dtype=torch.int32),
+            "seq_len": seq_len,
+        }
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/inference.py]: This issue has been resolved .

File: veomni/models/diffusers/minimax_h3/minimax_h3_core/minimax_h3_dit.py [hunk 14]
```diff
@@ -0,0 +1,536 @@
...
+    bounds = cu_seqlens.tolist()
+    for start, stop in zip(bounds[:-1], bounds[1:]):
+        if stop == start:
+            continue
+        seg_q = q[start:stop].transpose(0, 1).unsqueeze(0)
+        seg_k = k[start:stop].transpose(0, 1).unsqueeze(0)
+        seg_v = v[start:stop].transpose(0, 1).unsqueeze(0)
+        seg_out = attention_forward(seg_q, seg_k, seg_v, scale=softmax_scale)
+        out[start:stop] = seg_out.squeeze(0).transpose(0, 1)
+    return out
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/minimax_h3_core/minimax_h3_dit.py:99]: This issue has been resolved.

File: veomni/models/diffusers/minimax_h3/minimax_h3_core/offline_loader.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+    def __iter__(self):
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info is None:
+            paths = list(self._paths)
+        else:
+            per_worker = len(self._paths) // worker_info.num_workers
+            wid = worker_info.id
+            start = wid * per_worker
+            end = start + per_worker if wid < worker_info.num_workers - 1 else len(self._paths)
+            paths = list(self._paths[start:end])
```
[the remark this answers is not in the document]
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/minimax_h3_core/offline_loader.py]: This issue has been resolved.

File: veomni/data/multimodal/dit/minimax_h3_preprocess.py (hunk 7 above)
Comment (vvyuervv) [reply veomni/data/multimodal/dit/minimax_h3_preprocess.py]: The data processing logic differs significantly. If changes are needed, in addition to the model initiated for integration this time, other interfaces also need to be modified. This can be handled as a separate task.

File: veomni/trainer/dit_trainer.py [hunk 16]
```diff
@@ -322,7 +336,12 @@ def _build_model_assets(self):
 
     def _build_data_transform(self):
         args: VeOmniDiTArguments = self.base.args
-        if self.training_task == "offline_training":
+        if self.training_task == "offline_embedding":
```
Comment (Coach257) [inline veomni/trainer/dit_trainer.py]: if offline_training: dit_offline; else: args.data.data_transform
offline_embedding & online_training use the same transform

File: veomni/models/diffusers/minimax_h3/trainer.py [hunk 17]
```diff
@@ -0,0 +1,170 @@
+"""MiniMax H3 training glue: DiTTrainer subclass + fp32-master optimizer.
```
Comment (Coach257) [inline veomni/models/diffusers/minimax_h3/trainer.py]: It seems this file could be deleted

File: veomni/data/multimodal/dit/minimax_h3_preprocess.py (hunk 7 above)
Comment (Coach257) [inline veomni/data/multimodal/dit/minimax_h3_preprocess.py]: Move to veomni/data/multimodal/dit/preprocess.py

File: veomni/trainer/dit_trainer.py (hunk 16 above)
Comment (vvyuervv) [reply veomni/trainer/dit_trainer.py]: Merged. The default value of args.data.data_transform is changed to dit_online.

File: veomni/data/multimodal/dit/minimax_h3_preprocess.py (hunk 7 above)
Comment (vvyuervv) [reply veomni/data/multimodal/dit/minimax_h3_preprocess.py]: done

File: veomni/models/diffusers/minimax_h3/trainer.py (hunk 17 above)
Comment (vvyuervv) [reply veomni/models/diffusers/minimax_h3/trainer.py]: done

File: veomni/trainer/dit_trainer.py [hunk 18]
```diff
@@ -236,7 +248,9 @@ def _setup(self):
         # (default), so it was set to 1. Recompute now that dyn_bsz=False.
         args.train.dataloader_batch_size = args.train.global_batch_size // get_parallel_state().dp_size
         if args.train.training_task == "offline_embedding":
-            assert args.data.datasets_type == "mapping", "Datasets type must be mapping for offline embedding."
+            assert args.data.datasets_type in ("mapping", "minimax_h3_online"), (
```
Comment (Coach257) [inline veomni/trainer/dit_trainer.py]: No need to change this. The  datasets_type is still mapping. `minimax_h3_online` is `data_transform`.

**Comment by Coach257** [comment]
TODO: sp support; ops support

**Review by Coach257 — APPROVED** [review]
